doc/user-guide/data_model.rst

```diff
@@ -0,0 +1,251 @@
+.. _pyvista_data_model:
+
+
+PyVista Data Model
+==================
+PyVista uses the same data types as VTK, but structures them in a more
+pythonic manner for ease of use. If you'd like a background for how
+VTK structures its data, see `Introduction to VTK in Python by Kitware
+<https://vimeo.com/32232190>`_, as well as the numerous code examples
+on `Kitware's GitHub site
+<https://kitware.github.io/vtk-examples/site/>`_. An excellent
+introduction to mathematical concept implemented in VTK is provided by
+the `Discrete Differential Geometry YouTube Series
+<https://www.youtube.com/playlist?list=PL9_jI1bdZmz0hIrNCMQW1YmZysAiIYSSS>`_
+by Prof. Keenan Crane at Carnegie Melon.
+
+
+PyVista DataSet
+===============
+
+At the most fundamental level, all PyVista geometry objects are
+:ref:`ref_dataset`.  A dataset is a surface or volume in 3D space
+containing points, cells, and attributes describing that geometry.
+
+Geometry in PyVista is represented as points and cells.  In certain
```

Comment:
A picture is worth a thousand words. Can we add:

![image](https://user-images.githubusercontent.com/77850046/128108464-4b2d6434-432c-4f00-a821-fd97e5356295.png)

Also, can we change this paragraph to:

"Geometry in PyVista is represented as points and cells. As a simple example, consider the following 5x5 array:

```python
values = np.array([
    [1, 0, 0, 1, 1],
    [0, 1, 0, 0, 1],
    [1, 1, 0, 0, 0],
    [1, 0, 1, 0, 1],
    [0, 1, 1, 1, 0]])
```

We clearly see this dataset consists only of 1's and 0's, but where are they positioned in space? If we assume the data lives in a `UniformGrid`, then the (x, y) positions can be given by the indeces of the array. For example, point (3, 4) (NOTE: arrays are 0-indexed) is 1 and point (2, 3) is 0. But what if these datapoints aren't all in the plane and aren't uniformly spaced apart? We would need a way to describe the position of each of these points in space, but we're limited to expressing the values themselves as we've done above (lists of arrays with indeces). `VTK` (and hence `pyvista`) have multiple classes that represent different data shapes. The most important dataset classes are shown below:

![image](https://user-images.githubusercontent.com/77850046/128110221-281abd42-a46b-483b-8f39-86790d9c5373.png)

Here, the datasets are listed from most to least complex. That is, every dataset can be represented as an `UnstructuredGrid`, but `UnstructuredGrid`s take the most amount of memory to store since they must account for every data point individually. On the other hand, since `vtkImageData` (`UnformGrid` data) is uniformly spaced, a few integers and floats can describe the shape, so it takes the least amount of memory to store."

Reply:
Added most of this in 8e39b1ab1e70df7a748368c1865c36616ad3fbb7 (including figures, build docs to see).

![image](https://user-images.githubusercontent.com/11981631/128662739-b05a7ae2-79cb-4f18-99de-e6770666ed42.png)

I didn't add the entirety of your suggestion, so I'm leaving this unresolved for your second pass.